### DISCUSSION TITLE
How to start BackgroundTask from code

### DISCUSSION THREAD

**Discussion by vitalybrandes** [OrchardCMS/OrchardCore#7253] [Q&A] [answered]
I saw this somewhere and cannot find :-(
How to start BackgroundTask from code?

**Comment by hishamco** [comment, marked as answer]
Simply you can create your task by implementing the `OrchardCore.BackgroundTasks.IBackgroundTask`

**Reply by vitalybrandes** [reply]
I did.
As well createed cronrule,
but i need to run it manually from the controller.

**Reply by hishamco** [reply]
I think you could call `DoWorkAsync()`

**Reply by vitalybrandes** [reply]
Do you know maybe where to find some sample code?

**Reply by hishamco** [reply]
Try to check Demo module, also try to search in the source, the `DoWorkAsync()` should be called somewhere

**Reply by vitalybrandes** [reply]
didnt find any demo :-(

**Reply by hishamco** [reply]
This is how it called in `ModularBackgroundService ` 

https://github.com/OrchardCMS/OrchardCore/blob/943f502e790ee847335443a2ccac45b7df13ab41/src/OrchardCore/OrchardCore/Modules/ModularBackgroundService.cs#L126

**Reply by vitalybrandes** [reply]
Thank you!

**Comment by dodyg** [comment]
How does one enable the background task without relying on admin interface? Is there anyway to enable it via code?

**Reply by deanmarcussen** [reply]
Just apply the background task attribute with a schedule @dodyg 

https://github.com/OrchardCMS/OrchardCore/blob/3e03ed26f3a75ca49885f3002e2ea7f8dfb50401/src/OrchardCore.Modules/OrchardCore.AuditTrail/Services/AuditTrailBackgroundTask.cs#L14-L17

(and register the service)

The ui is mostly for editing the schedule, but if you are doing it by code, no need.

**Reply by dodyg** [reply]
Thanks!

**Reply by dodyg** [reply]
Hmm..is there anything else I need to enable from code? I created a dummy background service and it never got executed.

![trigger](https://user-images.githubusercontent.com/1761249/138851977-06bd2a32-173f-4c25-b85a-6b6eb726e6f3.png)
![trigger-2](https://user-images.githubusercontent.com/1761249/138851982-70e93180-ae3f-4190-b4e5-afa2edfb8f05.png)

This is the host Startup.cs

![host](https://user-images.githubusercontent.com/1761249/138853397-f0835c9c-2989-4cbd-817e-c386ce7585f9.png)

**Reply by dodyg** [reply]
The problem is if I change the host Startup.cs to 

``` csharp
        public void ConfigureServices(IServiceCollection services)
        {
            Conf.Emails(services, _config);
            services.AddMemoryCache();
            services.AddOrchardCore(x =>
           {
               x.AddBackgroundService();
           }).AddMvc();
            services.AddImageSharp();
        }
```

The module stops working without showing any exception.

**Reply by dodyg** [reply]
Ok this works

``` csharp

        public void ConfigureServices(IServiceCollection services)
        {
            Conf.Emails(services, _config);
            services.AddMemoryCache();
            services.AddOrchardCore()
                .AddBackgroundService()
                .AddMvc();

            services.AddImageSharp();
        }
```
